src/main/java/org/apache/sysml/runtime/matrix/data/SparseBlockCOO.java

```diff
@@ -194,7 +194,46 @@ public boolean isAllocated(int r) {
 
 	@Override
 	public boolean checkValidity(int rlen, int clen, long nnz, boolean strict) {
-		//empty implementation
+		//1. correct meta data
+		if(rlen < 0 || clen < 0) {
+			throw new RuntimeException("Invalid block dimensions: "+rlen+" "+clen);
+		}
+
+		//2. correct array lengths
+		if(_size != nnz && _cindexes.length < nnz && _rindexes.length < nnz && _values.length < nnz) {
+			throw new RuntimeException("Incorrect array lengths.");
+		}
+
+		//3. sorted values wrt to col indexes wrt to a given row index
+		for( int i=0; i<rlen; i++ ) {
+			int apos = pos(i);
+			int alen = size(i);
+			for(int k=apos+i; k<apos+alen; k++)
+				if( _cindexes[k+1] >= _cindexes[k] )
+					throw new RuntimeException("Wrong sparse row ordering: "
+							+ k + " "+_cindexes[k-1]+" "+_cindexes[k]);
+			for( int k=apos; k<apos+alen; k++ )
+				if(_values[k] == 0)
+					throw new RuntimeException("Wrong sparse row: zero at "
+							+ k + " at col index " + _cindexes[k]);
+			if(_rindexes[i+1] < _rindexes[i])
+				throw new RuntimeException("Wrong sorted order of row indices");
```

Comment:
In contrast to CSR, the `_rindexes` array in COO does not refer to row pointers but the actual row indexes. Hence, this check requires a separate loop over nnz not rlen. Otherwise it would either not consider all row indexes, or run out of bounds. Also the indexed access with i+1 could run out of bounds.